src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java

```diff
@@ -0,0 +1,118 @@
+
+/**
+ * Normally javac invokes annotation processors, but when the scala or kapt plugin are installed they will want to invoke
+ * the annotation processor themselves.
+ * See Java and Kapt implementations for a more deep understanding of the things passed by the children.
+ */
+public abstract class AnnotationProcessorInvoker<T extends Task> {
+	protected final Project project;
+	private final Collection<Configuration> annotationProcessorConfigurations;
+	protected final TaskCollection<T> invokerTasks;
+
+	protected AnnotationProcessorInvoker(Project project,
+										Collection<Configuration> annotationProcessorConfigurations,
+										TaskCollection<T> invokerTasks) {
+		this.project = project;
+		this.annotationProcessorConfigurations = annotationProcessorConfigurations;
+		this.invokerTasks = invokerTasks;
+	}
+
+	protected abstract void passArgument(T compileTask, String key, String value);
+
+	protected abstract File getDestinationDir(T task);
+
+	protected final String getRefmapDestination(T task, LoomGradleExtension extension) throws IOException {
+		return new File(getDestinationDir(task), extension.getRefmapName()).getCanonicalPath();
+	}
+
+	private void passMixinArguments(T task) {
+		try {
+			LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
+			Map<String, String> args = new HashMap<String, String>() {{
+					put("inMapFileNamedIntermediary", extension.getMappingsProvider().tinyMappings.getCanonicalPath());
+					put("outMapFileNamedIntermediary", extension.getMappingsProvider().mappingsMixinExport.getCanonicalPath());
+					put("outRefMapFile", getRefmapDestination(task, extension));
+					put("defaultObfuscationEnv", "named:intermediary");
+				}};
+
+			project.getLogger().info("Outputting refmap to dir: " + getDestinationDir(task) + " for compile task: " + task);
+			args.forEach((k, v) -> passArgument(task, k, v));
+		} catch (IOException e) {
+			project.getLogger().error("Could not configure mixin annotation processors", e);
+		}
+	}
+
+	public void configureMixin() {
+		ConfigurationContainer configs = project.getConfigurations();
+		LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
+
+		if (!extension.ideSync()) {
+			for (Configuration processorConfig : annotationProcessorConfigurations) {
+				project.getLogger().info("Adding mixin to classpath of AP config: " + processorConfig.getName());
+				// Pass named MC classpath to mixin AP classpath
+				processorConfig.extendsFrom(
+								configs.getByName(Constants.MINECRAFT_NAMED),
+								configs.getByName(Constants.MOD_COMPILE_CLASSPATH_MAPPED),
+								configs.getByName(Constants.MAPPINGS_FINAL)
+				);
+
+				// Add Mixin and mixin extensions (fabric-mixin-compile-extensions pulls mixin itself too)
+				project.getDependencies().add(processorConfig.getName(),
+								"net.fabricmc:fabric-mixin-compile-extensions:" + Constants.MIXIN_COMPILE_EXTENSIONS_VERSION);
```

Comment:
I know im still against kotlin mixins, but is this really needed on the kotlin annotation processor? Does kapt work without this.

Reply:
When kapt it applied it invokes java mixins instead of the java plugin, so it needs to have mixin and the mc stuff on its classpath

Reply:
Ok, thats what I assumed.